### PR TITLE
Decouple Security from react-router

### PR DESCRIPTION
**Pull request by denysoblohin-okta**
Internal ref: https://oktainc.atlassian.net/browse/OKTA-333775
Resolves: https://github.com/okta/okta-oidc-js/issues/910

### Changes:
- Added required prop `restoreOriginalUri` to `Security`
- Updated readme with setting `restoreOriginalUri` callback for `react-router`

### CODE REVIEW

File: test/jest/secureRoute.test.tsx
Comment (denysoblohin-okta) [inline test/jest/secureRoute.test.tsx:309]: To prevent this:
```
ts-jest[ts-compiler] (WARN) test/jest/secureRoute.test.tsx:308:46 - error TS2339: Property 'exact' does not exist on type 'Readonly<{}> & Readonly<{ children?: ReactNode; }>'.
308       expect(secureRoute.find(Route).props().exact).toBe(true);
                                                 ~~~~~
test/jest/secureRoute.test.tsx:325:46 - error TS2339: Property 'strict' does not exist on type 'Readonly<{}> & Readonly<{ children?: ReactNode; }>'.
325       expect(secureRoute.find(Route).props().strict).toBe(true);
                                                 ~~~~~~
test/jest/secureRoute.test.tsx:342:46 - error TS2339: Property 'sensitive' does not exist on type 'Readonly<{}> & Readonly<{ children?: ReactNode; }>'.
342       expect(secureRoute.find(Route).props().sensitive).toBe(true);
```

**Comment by denysoblohin-okta** [comment]
I think there is no need to add prop `restoreOriginalUri` to `Security` as we can configure it in `oktaAuth` which is passed to `Security`

File: CHANGELOG.md
Comment (shuowu) [inline CHANGELOG.md:5]: nit: adding -> ~adding~

Comment (shuowu) [reply CHANGELOG.md:5]: The default `restoreOriginalUri ` from `auth-js` changes `window.location`, which reload the whole SPA, we should also accept an `restoreOriginalUri ` (required) callback from `<Security>` prop to override the default one from auth-js.

File: src/Security.tsx
Comment (shuowu) [inline src/Security.tsx:26]: accept `restoreOriginalUri ` here to avoid reloading the page during after login redirect.

File: src/Security.tsx
Comment (shuowu) [inline src/Security.tsx:47]: As users should have control on configs of oktaAuth instance, we can do check here and provide proper warning if two restoreOriginalUri are provided (one from Security prop, one from oktaAuth options).

**Review by swiftone — COMMENTED** [review]
LGTM after addressing Shuo's comments

**Comment by shuowu** [comment]
> I think there is no need to add prop `restoreOriginalUri` to `Security` as we can configure it in `oktaAuth` which is passed to `Security`

The purpose of a `restoreOriginalUri` prop is to force users to provide their own in-app redirect logic. If we only leave it to `auth-js`, most react apps will end up with the default behavior that reloads the page.

**Comment by denysoblohin-okta** [comment]
@shuowu Updated PR with required prop `restoreOriginalUri `

File: CHANGELOG.md
Comment (shuowu) [inline CHANGELOG.md:5]: added -> adds

let's also mention what has been changed to the default behavior here.

File: README.md
Comment (shuowu) [inline README.md:130]: let's use `App` to keep sample code consistent.

File: README.md
Comment (shuowu) [inline README.md:505]: should this change belong to `Migrating from 4.x to 5.x` ?

File: README.md
Comment (shuowu) [inline README.md:486]: nit: have required -> explicitly requires

also, move the code snippet to this section as mentioned in the comment below.

File: src/Security.tsx
Comment (shuowu) [inline src/Security.tsx:22]: nit: move the required prop above the optional props.

File: src/Security.tsx
Comment (shuowu) [reply src/Security.tsx:47]: @denysoblohin-okta ^ thoughts on the comment above?

Comment (denysoblohin-okta) [reply src/Security.tsx:47]: @shuowu Updated PR

File: README.md
Comment (shuowu) [inline README.md:174]: move oktaAuth initialization out of the `App` functional component to avoid multiple instances be initialized.

File: README.md
Comment (shuowu) [inline README.md:484]: it would be great to also explain why Security comp requires this prop, like `to decouple from react-router`.

File: src/Security.tsx
Comment (shuowu) [inline src/Security.tsx:48]: nit: provied -> are provided

How about 
```
Two custom restoreOriginalUri callbacks are detected. The one from the OktaAuth configuration will be overridden by the provided restoreOriginalUri prop from the Security component.
```

**Review by shuowu — APPROVED** [review]

**Pull request merged**
